### ISSUE TITLE
Clang rejects use of static constexpr member function in noexcept complete-class context 

### ISSUE THREAD

**Issue by ranaanoop** [llvm/llvm-project#58245]
The following valid program is rejected by clang as well as other compilers. [Demo](https://godbolt.org/z/Ph4vfGjzd)

```
struct A {
   constexpr static bool func() 
   { 
       return true; 
   }
   //--------------vvvvvv------->works as expected   #1
   void f(bool V1 = func())
   { 
      bool V2 = func(); //works as expected          #2
   }
   //-----------------vvvvvv---->DOESN'T WORK?       #3
   void g()  noexcept(func()) 
   {
      ;
   }

};
```
From [class.mem.general](https://eel.is/c++draft/class.mem.general#7.4):

> A complete-class context of a class is a
> * function body
> * default argument
> * **noexcept specifier**

As we can see the noexcept speficier is included in the complete-class context so the `#3` should work.

**Comment by shafik** [comment]
Note this version is accepted by those compilers: https://godbolt.org/z/78cKKj1W5

```cpp
struct A {
   constexpr static bool func();
   void g()  noexcept(noexcept(func())) ;
};

constexpr  bool A::func() {  return true; }

void A::g()  noexcept(noexcept(func())) {}

```

**Comment by shafik** [comment]
Also note the following which has a default member initializer should be consider in the complete context but is not accepted by any of the three compilers either: https://godbolt.org/z/4WEcrjMcz

```cpp
struct A {
  int static consteval f() { return 1;}
  int arr[1] = {f()}; // Not ok
};

int main() {
    int arr[A::f()] = {A::f()}; // Ok
}
```

**Comment by ranaanoop** [comment]
@shafik Yes, I'm aware that `struct A {constexpr static bool func();void g()  noexcept(noexcept(func())) ;};` will work but my point is that according to the current wording `void g()  noexcept(func()) {}` should also work.

**Comment by shafik** [comment]
I wonder if this is related to [defect report 2510](https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#2510).

Interesting to note that [Revision 4 of Declarations and where to find them](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r4.html) removed noexcept from the list of complete class contexts but it was added back in [revision 5 but no explanation in the revision history explain it](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r5.html).
